Cabal/src/Distribution/Simple/PreProcess/Types.hs

```diff
@@ -0,0 +1,109 @@
+--  >          stuff <- readFile inFile
+--  >          writeFile outFile ("-- preprocessed as a test\n\n" ++ stuff)
+--  >          return ExitSuccess
+--
+--  We split the input and output file names into a base directory and the
+--  rest of the file name. The input base dir is the path in the list of search
+--  dirs that this file was found in. The output base dir is the build dir where
+--  all the generated source files are put.
+--
+--  The reason for splitting it up this way is that some pre-processors don't
+--  simply generate one output .hs file from one input file but have
+--  dependencies on other generated files (notably c2hs, where building one
+--  .hs file may require reading other .chi files, and then compiling the .hs
+--  file may require reading a generated .h file). In these cases the generated
+--  files need to embed relative path names to each other (eg the generated .hs
+--  file mentions the .h file in the FFI imports). This path must be relative to
+--  the base directory where the generated files are located, it cannot be
+--  relative to the top level of the build tree because the compilers do not
+--  look for .h files relative to there, ie we do not use \"-I .\", instead we
+--  use \"-I dist\/build\" (or whatever dist dir has been set by the user)
+--
+--  Most pre-processors do not care of course, so mkSimplePreProcessor and
+--  runSimplePreProcessor functions handle the simple case.
+data PreProcessor = PreProcessor
+  { -- Is the output of the pre-processor platform independent? eg happy output
+    -- is portable haskell but c2hs's output is platform dependent.
+    -- This matters since only platform independent generated code can be
+    -- included into a source tarball.
+    platformIndependent :: Bool
+  , -- TODO: deal with pre-processors that have implementation dependent output
+    --       eg alex and happy have --ghc flags. However we can't really include
+    --       ghc-specific code into supposedly portable source tarballs.
+
+    ppOrdering
+      :: Verbosity
+      -> [FilePath] -- Source directories
+      -> [ModuleName] -- Module names
+      -> IO [ModuleName] -- Sorted modules
+
+  -- ^ This function can reorder /all/ modules, not just those that the
+  -- require the preprocessor in question. As such, this function should be
+  -- well-behaved and not reorder modules it doesn't have dominion over!
+  --
+  -- @since 3.8.1.0
+  , runPreProcessor
+      :: (FilePath, FilePath) -- Location of the source file relative to a base dir
+      -> (FilePath, FilePath) -- Output file name, relative to an output base dir
+      -> Verbosity -- verbosity
+      -> IO () -- Should exit if the preprocessor fails
+  }
+
+-- | A suffix (or file extension).
+--
+-- Mostly used to decide which preprocessor to use, e.g. files with suffix @"y"@
+-- are usually processed by the @"happy"@ build tool.
+newtype Suffix = Suffix String
+  deriving (Eq, Ord, Show, Generic, IsString)
```

Comment:
Looking at CI, it looks like the stock Show instance of Suffix is used to print human-readable messages, which is not what it should be used for. Either Show should be derived `via String` or another function that unwraps it should be used to build this message:  `Could not find module: MyLibHelperModule with any suffix: ["gc","chs","hsc","x","y","ly","cpphs","hs","lhs","hsig","lhsig"].`

Reply:
Use
```
instance Pretty Suffix where
  pretty (Suffix s) = Disp.text s
```
and change `Cabal/src/Distribution/Simple/Errors.hs` to do
```
      ++ intercalate "," (map prettyShow suffixes)
```

Reply:
Indeed. Fixing...